pkg/buses/bus.go

```diff
@@ -0,0 +1,166 @@
+ * limitations under the License.
+ */
+
+package buses
+
+import (
+	"fmt"
+
+	channelsv1alpha1 "github.com/knative/eventing/pkg/apis/channels/v1alpha1"
+)
+
+type bus struct {
+	busRef       BusReference
+	handlerFuncs EventHandlerFuncs
+
+	reconciler *Reconciler
+	cache      *Cache
+	dispatcher *MessageDispatcher
+	receiver   *MessageReceiver
+}
+
+// BusOpts holds configuration options for new buses. These options are not
+// required for proper operation of the bus, but are useful to override the
+// default behavior and for testing.
+type BusOpts struct {
+	// MasterURL is the address of the Kubernetes API server. Overrides any
+	// value in kubeconfig. Only required if out-of-cluster.
+	MasterURL string
+	// KubeConfig is the path to a kubeconfig. Only required if out-of-cluster.
+	KubeConfig string
+
+	// Cache to use for this bus. A cache will be created for the bus if not
+	// specified.
+	Cache *Cache
+	// Reconciler to use for this bus. A reconciler wil be created for the bus
+	// if not specified.
+	Reconciler *Reconciler
+	// MessageDispatcher to use for this bus. The message dispatcher is used to
+	// send messages from the bus to a subscriber. A message dispatcher will be
+	// created for the bus if needed and not specified.
+	MessageDispatcher *MessageDispatcher
+	// MessageReceiver to use for this bus. The message receiver is used to
+	// receive message sent to a channel. A message receiver will be created
+	// for the bus if needed and not specified.
+	MessageReceiver *MessageReceiver
+}
+
+// BusProvisioner provisions channels and subscriptions for a bus on backing
+// infrastructure.
+type BusProvisioner interface {
+	Run(threadiness int, stopCh <-chan struct{})
+}
+
+// NewBusProvisioner creates a new provisioner for a specific bus.
+// EventHandlerFuncs are used to be notified when a channel or subscription is
+// created, updated or removed.
+func NewBusProvisioner(busRef BusReference, handlerFuncs EventHandlerFuncs, opts *BusOpts) BusProvisioner {
+	bus := &bus{
+		busRef:       busRef,
+		handlerFuncs: handlerFuncs,
+	}
+
+	if opts.Cache == nil {
+		opts.Cache = NewCache()
+	}
+	bus.cache = opts.Cache
+
+	if opts.Reconciler == nil {
+		opts.Reconciler = NewReconciler(Provisioner, opts.MasterURL, opts.KubeConfig, bus.cache, handlerFuncs)
+	}
+	bus.reconciler = opts.Reconciler
+
+	return bus
+}
+
+// BusDispatcher dispatches messages from channels to subscribers via backing
+// infrastructure.
+type BusDispatcher interface {
+	Run(threadiness int, stopCh <-chan struct{})
+	DispatchMessage(subscriptionRef SubscriptionReference, message *Message) error
+}
+
+// NewBusDispatcher creates a new dispatcher for a specific bus.
+// EventHandlerFuncs are used to be notified when a subscription is created,
+// updated or removed, or a message is received.
+func NewBusDispatcher(busRef BusReference, handlerFuncs EventHandlerFuncs, opts *BusOpts) BusDispatcher {
+	bus := &bus{
```

Comment:
will do, although it's not quite as clean as you'd sugest because there's a cyclic dependency between the bus and the receiver I still need to untangle.